{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "import seaborn as sns\n",
    "import os\n",
    "import re\n",
    "from scipy import stats\n",
    "import gc\n",
    "np.random.seed(12)\n",
    "import matplotlib.pyplot as plt\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filterParquet(path):\n",
    "\t# Note, will keep the chrom column on the dataframe\n",
    "\tdf = pd.read_parquet(path)\n",
    "\n",
    "\t# Shift methylation to between 0 and 100.\n",
    "\tdf = df * 100\n",
    "\n",
    "\tassert df.isnull().sum().sum() == 0, \"There are missing values in the dataframe\"\n",
    "\n",
    "\t# Add a chromosome column\n",
    "\tdf['chrom'] = df.index.str.split(':').str[0]\n",
    "\n",
    "\t# Drop the Sex chromosomes\n",
    "\tdf = df[df['chrom'].str.match(pat=r'[0-9]{1,2}')]\n",
    "\t\n",
    "\treturn df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "angus_all = filterParquet('./aligned2Angus/All.samples.meth.10X.parquet')\n",
    "brahman_all = filterParquet('./aligned2Brahman/All.samples.meth.10X.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample: AxA_F103\n",
      "Quantification bias: 2.3603473613008297\n",
      "Sample: AxA_F105\n",
      "Quantification bias: 2.3595326171354696\n",
      "Sample: AxA_F52\n",
      "Quantification bias: 2.6271052784220776\n",
      "Sample: AxA_F53\n",
      "Quantification bias: 2.366704981203307\n",
      "Sample: AxA_F60\n",
      "Quantification bias: 2.3231867315550794\n",
      "Sample: AxA_F7\n",
      "Quantification bias: 2.3399440500307103\n",
      "Sample: AxB_F100\n",
      "Quantification bias: 1.4773883072076308\n",
      "Sample: AxB_F104\n",
      "Quantification bias: 1.1128804587393681\n",
      "Sample: AxB_F106\n",
      "Quantification bias: 1.0850219367152938\n",
      "Sample: AxB_F61\n",
      "Quantification bias: 1.389872680025008\n",
      "Sample: AxB_F74\n",
      "Quantification bias: 1.121484169765189\n",
      "Sample: AxB_F97\n",
      "Quantification bias: 0.01143455834836027\n",
      "Sample: BxA_F13\n",
      "Quantification bias: 0.8705831345916241\n",
      "Sample: BxA_F62\n",
      "Quantification bias: 0.834458390535048\n",
      "Sample: BxA_F77\n",
      "Quantification bias: 0.963108923837918\n",
      "Sample: BxA_F80\n",
      "Quantification bias: 0.9675014492105982\n",
      "Sample: BxA_F8\n",
      "Quantification bias: 0.93452920035478\n",
      "Sample: BxA_F91\n",
      "Quantification bias: 0.9040434558612929\n",
      "Sample: BxB_F22\n",
      "Quantification bias: 0.437065292594787\n",
      "Sample: BxB_F46\n",
      "Quantification bias: 0.3716013676859283\n",
      "Sample: BxB_F56\n",
      "Quantification bias: 0.3931815915893837\n",
      "Sample: BxB_F65\n",
      "Quantification bias: 0.5286218888844211\n",
      "Sample: BxB_F78\n",
      "Quantification bias: 0.45984981928245\n",
      "Sample: BxB_F99\n",
      "Quantification bias: 0.4256472692459736\n"
     ]
    }
   ],
   "source": [
    "# Get the methylation of every site in the parquet file and\n",
    "# for each sample and compare, at the sample level if there is a\n",
    "# significant difference in the methylation distribution\n",
    "\n",
    "p_values = {sample: [] for sample in angus_all.columns[:-1]}\n",
    "\n",
    "means = {'sampleID':[],'group':[],\n",
    "\t\t 'n_CpGs':[],\n",
    "\t\t 'angus_methylation_mean':[],\n",
    "\t\t 'brahman_methylation_mean':[],\n",
    "\t\t 'angus_methylation_std':[],\n",
    "\t\t 'brahman_methylation_std':[],\n",
    "\t\t 'difference':[],\n",
    "\t\t 'p_value':[],\n",
    "\t\t 'adjusted_p_value':[]}\n",
    "\n",
    "\n",
    "\n",
    "for sample in angus_all.columns[:-1]:\n",
    "\n",
    "\tgroup = sample.split(\"_\")[0]\n",
    "\tsampleID = sample.split(\"_\")[1]\n",
    "\n",
    "\t# Create an index of each genome for any necessary\n",
    "\t# downsampling\n",
    "\tangus_all_idx = np.arange(angus_all.shape[0])\n",
    "\tbrahman_all_idx = np.arange(brahman_all.shape[0])\n",
    "\tif len(angus_all_idx) > len(brahman_all_idx):\n",
    "\t\tangus_all_idx = np.random.choice(angus_all_idx, len(brahman_all_idx), replace=False)\n",
    "\telif len(angus_all_idx) < len(brahman_all_idx):\n",
    "\t\tbrahman_all_idx = np.random.choice(brahman_all_idx, len(angus_all_idx), replace=False)\n",
    "\n",
    "\tmapped_2_angus_methylation = angus_all[sample].values\n",
    "\tmapped_2_brahman_methylation = brahman_all[sample].values\n",
    "\n",
    "\t# Downsample the methylation arrays\n",
    "\tmapped_2_angus_methylation = mapped_2_angus_methylation[angus_all_idx]\n",
    "\tmapped_2_brahman_methylation = mapped_2_brahman_methylation[brahman_all_idx]\n",
    "\n",
    "\t# Compute the mean and median methylation for each sample\n",
    "\t# and genome\n",
    "\tmeans['sampleID'].append(sampleID)\n",
    "\tmeans['n_CpGs'].append(len(mapped_2_angus_methylation))\n",
    "\tmeans['angus_methylation_mean'].append(mapped_2_angus_methylation.mean())\n",
    "\tmeans['angus_methylation_std'].append(np.std(mapped_2_angus_methylation))\n",
    "\t#means['reference'].append('Angus')\n",
    "\tmeans['group'].append(group)\n",
    "\t\n",
    "\t#means['sampleID'].append(sampleID)\n",
    "\tmeans['brahman_methylation_mean'].append(mapped_2_brahman_methylation.mean())\n",
    "\tmeans['brahman_methylation_std'].append(np.std(mapped_2_brahman_methylation))\n",
    "\t#means['reference'].append('Brahman')\n",
    "\t#means['group'].append(group)\n",
    "\n",
    "\n",
    "\t# Compare the distributions of each sample\n",
    "\t_, pvalue = stats.wilcoxon(mapped_2_angus_methylation, mapped_2_brahman_methylation)\n",
    "\tp_values[sample].append(pvalue)\n",
    "\tmeans['p_value'].append(pvalue)\n",
    "\n",
    "\t# Get the percentage difference between the means\n",
    "\t# of the two distributions\n",
    "\t#percentage_diff = ((mapped_2_angus_methylation.mean() - mapped_2_brahman_methylation.mean()) / (mapped_2_angus_methylation.mean() + mapped_2_brahman_methylation.mean()/2)) * 100\n",
    "\t#difference = np.mean((np.abs(mapped_2_angus_methylation - mapped_2_brahman_methylation)))\n",
    "\tprint(f\"Sample: {sample}\\nQuantification bias: {np.abs(mapped_2_angus_methylation.mean() - mapped_2_brahman_methylation.mean())}\")\n",
    "\tdifference = np.abs(mapped_2_angus_methylation.mean() - mapped_2_brahman_methylation.mean())\n",
    "\t#means['difference'].append(np.abs(percentage_diff))\n",
    "\tmeans['difference'].append(difference)\n",
    "\t#print(f\"Sample: {sample}, p-value: {pvalue}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test the p-values for multiple testing\n",
    "multipletests(pd.DataFrame(p_values).values.flatten(), alpha=0.01, method='fdr_bh')\n",
    "means['adjusted_p_value'].extend(multipletests(pd.DataFrame(p_values).values.flatten(), alpha=0.01, method='fdr_bh')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(means).to_csv('/Users/callummacphillamy/Library/CloudStorage/OneDrive-UniversityofAdelaide/PhD/methylation_chapter/GigaScience_submission/supplementary_tables/all_CpGs_sample-wise.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 24/24 [05:05<00:00, 12.72s/it]\n"
     ]
    }
   ],
   "source": [
    "# S table re. Figure3A variance and what not\n",
    "all_cpgs = {'AxA':{'angus_ref_methylation':[],\n",
    "\t\t\t\t   'brahman_ref_methylation':[]},\n",
    "\t\t\t'AxB':{'angus_ref_methylation':[],\n",
    "\t\t\t\t   'brahman_ref_methylation':[]},\n",
    "\t\t\t'BxA':{'angus_ref_methylation':[],\n",
    "\t\t           'brahman_ref_methylation':[]},\n",
    "\t\t\t'BxB':{'angus_ref_methylation':[],\n",
    "\t\t           'brahman_ref_methylation':[]}}\n",
    "\n",
    "stable_dict = {'sampleID':[],\n",
    "\t\t\t   'n_CpGs':[],\n",
    "\t\t\t   'angus_methylation_mean':[],\n",
    "\t\t\t   'brahman_methylation_mean':[],\n",
    "\t\t\t   'angus_methylation_std_dev':[],\n",
    "\t\t\t   'brahman_methylation_std_dev':[],\n",
    "\t\t\t   'difference':[],\n",
    "\t\t\t   'p_value':[],\n",
    "\t\t\t   'group':[]}\n",
    "for sample in tqdm(angus_all.columns[:-1]):\n",
    "\n",
    "\tgroup = sample.split(\"_\")[0]\n",
    "\tsampleID = sample.split(\"_\")[1]\n",
    "\n",
    "\t# Create an index of each genome for any necessary\n",
    "\t# downsampling\n",
    "\tangus_all_idx = np.arange(angus_all.shape[0])\n",
    "\tbrahman_all_idx = np.arange(brahman_all.shape[0])\n",
    "\tif len(angus_all_idx) > len(brahman_all_idx):\n",
    "\t\tangus_all_idx = np.random.choice(angus_all_idx, len(brahman_all_idx), replace=False)\n",
    "\telif len(angus_all_idx) < len(brahman_all_idx):\n",
    "\t\tbrahman_all_idx = np.random.choice(brahman_all_idx, len(angus_all_idx), replace=False)\n",
    "\n",
    "\tmapped_2_angus_methylation = angus_all[sample].values\n",
    "\tmapped_2_brahman_methylation = brahman_all[sample].values\n",
    "\n",
    "\t# Downsample the methylation arrays\n",
    "\tmapped_2_angus_methylation = mapped_2_angus_methylation[angus_all_idx]\n",
    "\tmapped_2_brahman_methylation = mapped_2_brahman_methylation[brahman_all_idx]\n",
    "\n",
    "\t# Compute the mean and median methylation for each sample\n",
    "\t# and genome\n",
    "\tstable_dict['sampleID'].append(sampleID)\n",
    "\tassert len(mapped_2_angus_methylation) == len(mapped_2_brahman_methylation), \"Downsampling failed\"\n",
    "\tstable_dict['n_CpGs'].append(len(mapped_2_angus_methylation))\n",
    "\tstable_dict['angus_methylation_mean'].append(mapped_2_angus_methylation.mean())\n",
    "\tstable_dict['brahman_methylation_mean'].append(mapped_2_brahman_methylation.mean())\n",
    "\tstable_dict['angus_methylation_std_dev'].append(np.std(mapped_2_angus_methylation))\n",
    "\tstable_dict['brahman_methylation_std_dev'].append(np.std(mapped_2_brahman_methylation))\n",
    "\tstable_dict['difference'].append(np.abs(mapped_2_angus_methylation.mean() - mapped_2_brahman_methylation.mean()))\n",
    "\tstable_dict['p_value'].append(stats.wilcoxon(mapped_2_angus_methylation, mapped_2_brahman_methylation).pvalue)\n",
    "\tstable_dict['group'].append(group)\n",
    "\t\n",
    "\t# stable_dict['sampleID'].append(sampleID)\n",
    "\t# stable_dict['methylation_mean'].append(mapped_2_brahman_methylation.mean())\n",
    "\t# stable_dict['methylation_std_dev'].append(np.std(mapped_2_brahman_methylation))\n",
    "\t# stable_dict['reference'].append('Brahman')\n",
    "\t# stable_dict['group'].append(group)\n",
    "\n",
    "\tall_cpgs[group]['angus_ref_methylation'].extend(mapped_2_angus_methylation)\n",
    "\tall_cpgs[group]['brahman_ref_methylation'].extend(mapped_2_brahman_methylation)\n",
    "#pd.DataFrame.from_dict(stable_dict).to_csv('/Users/callummacphillamy/Library/CloudStorage/OneDrive-UniversityofAdelaide/PhD/methylation_chapter/GigaScience_submission/supplementary_tables/all_CpGs_per_sample.csv', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[37], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m group \u001b[38;5;129;01min\u001b[39;00m all_cpgs\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[0;32m----> 2\u001b[0m \t\u001b[38;5;28mprint\u001b[39m(\u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mDataFrame\u001b[49m\u001b[43m(\u001b[49m\u001b[43mall_cpgs\u001b[49m\u001b[43m[\u001b[49m\u001b[43mgroup\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m      3\u001b[0m \t\u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/mambaforge/envs/wgbs/lib/python3.10/site-packages/pandas/core/frame.py:767\u001b[0m, in \u001b[0;36mDataFrame.__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    761\u001b[0m     mgr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_init_mgr(\n\u001b[1;32m    762\u001b[0m         data, axes\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m\"\u001b[39m: index, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcolumns\u001b[39m\u001b[38;5;124m\"\u001b[39m: columns}, dtype\u001b[38;5;241m=\u001b[39mdtype, copy\u001b[38;5;241m=\u001b[39mcopy\n\u001b[1;32m    763\u001b[0m     )\n\u001b[1;32m    765\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, \u001b[38;5;28mdict\u001b[39m):\n\u001b[1;32m    766\u001b[0m     \u001b[38;5;66;03m# GH#38939 de facto copy defaults to False only in non-dict cases\u001b[39;00m\n\u001b[0;32m--> 767\u001b[0m     mgr \u001b[38;5;241m=\u001b[39m \u001b[43mdict_to_mgr\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtyp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmanager\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    768\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, ma\u001b[38;5;241m.\u001b[39mMaskedArray):\n\u001b[1;32m    769\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mma\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m mrecords\n",
      "File \u001b[0;32m~/mambaforge/envs/wgbs/lib/python3.10/site-packages/pandas/core/internals/construction.py:503\u001b[0m, in \u001b[0;36mdict_to_mgr\u001b[0;34m(data, index, columns, dtype, typ, copy)\u001b[0m\n\u001b[1;32m    499\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    500\u001b[0m         \u001b[38;5;66;03m# dtype check to exclude e.g. range objects, scalars\u001b[39;00m\n\u001b[1;32m    501\u001b[0m         arrays \u001b[38;5;241m=\u001b[39m [x\u001b[38;5;241m.\u001b[39mcopy() \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(x, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m x \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m arrays]\n\u001b[0;32m--> 503\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43marrays_to_mgr\u001b[49m\u001b[43m(\u001b[49m\u001b[43marrays\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtyp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtyp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconsolidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/mambaforge/envs/wgbs/lib/python3.10/site-packages/pandas/core/internals/construction.py:119\u001b[0m, in \u001b[0;36marrays_to_mgr\u001b[0;34m(arrays, columns, index, dtype, verify_integrity, typ, consolidate)\u001b[0m\n\u001b[1;32m    116\u001b[0m         index \u001b[38;5;241m=\u001b[39m ensure_index(index)\n\u001b[1;32m    118\u001b[0m     \u001b[38;5;66;03m# don't force copy because getting jammed in an ndarray anyway\u001b[39;00m\n\u001b[0;32m--> 119\u001b[0m     arrays, refs \u001b[38;5;241m=\u001b[39m \u001b[43m_homogenize\u001b[49m\u001b[43m(\u001b[49m\u001b[43marrays\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    120\u001b[0m     \u001b[38;5;66;03m# _homogenize ensures\u001b[39;00m\n\u001b[1;32m    121\u001b[0m     \u001b[38;5;66;03m#  - all(len(x) == len(index) for x in arrays)\u001b[39;00m\n\u001b[1;32m    122\u001b[0m     \u001b[38;5;66;03m#  - all(x.ndim == 1 for x in arrays)\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    125\u001b[0m \n\u001b[1;32m    126\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    127\u001b[0m     index \u001b[38;5;241m=\u001b[39m ensure_index(index)\n",
      "File \u001b[0;32m~/mambaforge/envs/wgbs/lib/python3.10/site-packages/pandas/core/internals/construction.py:629\u001b[0m, in \u001b[0;36m_homogenize\u001b[0;34m(data, index, dtype)\u001b[0m\n\u001b[1;32m    626\u001b[0m         val \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(val)\n\u001b[1;32m    627\u001b[0m     val \u001b[38;5;241m=\u001b[39m lib\u001b[38;5;241m.\u001b[39mfast_multiget(val, oindex\u001b[38;5;241m.\u001b[39m_values, default\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mnan)\n\u001b[0;32m--> 629\u001b[0m val \u001b[38;5;241m=\u001b[39m \u001b[43msanitize_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43mval\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m    630\u001b[0m com\u001b[38;5;241m.\u001b[39mrequire_length_match(val, index)\n\u001b[1;32m    631\u001b[0m refs\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;28;01mNone\u001b[39;00m)\n",
      "File \u001b[0;32m~/mambaforge/envs/wgbs/lib/python3.10/site-packages/pandas/core/construction.py:651\u001b[0m, in \u001b[0;36msanitize_array\u001b[0;34m(data, index, dtype, copy, allow_2d)\u001b[0m\n\u001b[1;32m    648\u001b[0m     subarr \u001b[38;5;241m=\u001b[39m _try_cast(data, dtype, copy)\n\u001b[1;32m    650\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 651\u001b[0m     subarr \u001b[38;5;241m=\u001b[39m \u001b[43mmaybe_convert_platform\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    652\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m subarr\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mobject\u001b[39m:\n\u001b[1;32m    653\u001b[0m         subarr \u001b[38;5;241m=\u001b[39m cast(np\u001b[38;5;241m.\u001b[39mndarray, subarr)\n",
      "File \u001b[0;32m~/mambaforge/envs/wgbs/lib/python3.10/site-packages/pandas/core/dtypes/cast.py:139\u001b[0m, in \u001b[0;36mmaybe_convert_platform\u001b[0;34m(values)\u001b[0m\n\u001b[1;32m    137\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m _dtype_obj:\n\u001b[1;32m    138\u001b[0m     arr \u001b[38;5;241m=\u001b[39m cast(np\u001b[38;5;241m.\u001b[39mndarray, arr)\n\u001b[0;32m--> 139\u001b[0m     arr \u001b[38;5;241m=\u001b[39m \u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmaybe_convert_objects\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    141\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m arr\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for group in all_cpgs.keys():\n",
    "\tprint(pd.DataFrame(all_cpgs[group]))\n",
    "\tbreak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>n_CpGs</th>\n",
       "      <th>angus_methylation_mean</th>\n",
       "      <th>brahman_methylation_mean</th>\n",
       "      <th>angus_methylation_std_dev</th>\n",
       "      <th>brahman_methylation_std_dev</th>\n",
       "      <th>difference</th>\n",
       "      <th>p_value</th>\n",
       "      <th>adjusted_p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AxA</td>\n",
       "      <td>128592426</td>\n",
       "      <td>55.647847</td>\n",
       "      <td>53.669505</td>\n",
       "      <td>24.385631</td>\n",
       "      <td>25.980464</td>\n",
       "      <td>1.978342</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AxB</td>\n",
       "      <td>128592426</td>\n",
       "      <td>52.184830</td>\n",
       "      <td>51.393670</td>\n",
       "      <td>23.500377</td>\n",
       "      <td>24.145755</td>\n",
       "      <td>0.791160</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BxA</td>\n",
       "      <td>128592426</td>\n",
       "      <td>50.443728</td>\n",
       "      <td>49.756729</td>\n",
       "      <td>23.384975</td>\n",
       "      <td>23.916931</td>\n",
       "      <td>0.686999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BxB</td>\n",
       "      <td>128592426</td>\n",
       "      <td>52.726276</td>\n",
       "      <td>53.070719</td>\n",
       "      <td>25.131800</td>\n",
       "      <td>24.898284</td>\n",
       "      <td>0.344443</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  group     n_CpGs  angus_methylation_mean  brahman_methylation_mean  \\\n",
       "0   AxA  128592426               55.647847                 53.669505   \n",
       "1   AxB  128592426               52.184830                 51.393670   \n",
       "2   BxA  128592426               50.443728                 49.756729   \n",
       "3   BxB  128592426               52.726276                 53.070719   \n",
       "\n",
       "   angus_methylation_std_dev  brahman_methylation_std_dev  difference  \\\n",
       "0                  24.385631                    25.980464    1.978342   \n",
       "1                  23.500377                    24.145755    0.791160   \n",
       "2                  23.384975                    23.916931    0.686999   \n",
       "3                  25.131800                    24.898284    0.344443   \n",
       "\n",
       "   p_value  adjusted_p_value  \n",
       "0      0.0               0.0  \n",
       "1      0.0               0.0  \n",
       "2      0.0               0.0  \n",
       "3      0.0               0.0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_p_vals = {'group':[],\n",
    "\t\t\t\t'n_CpGs':[],\n",
    "\t\t\t\t'angus_methylation_mean':[],\n",
    "\t\t\t\t'brahman_methylation_mean':[],\n",
    "\t\t\t\t'angus_methylation_std_dev':[],\n",
    "\t\t\t\t'brahman_methylation_std_dev':[],\n",
    "\t\t\t\t'difference':[],\n",
    "\t\t\t\t'p_value':[],\n",
    "\t\t\t\t'adjusted_p_value':[]}\n",
    "pvals = []\n",
    "\n",
    "for group in all_cpgs.keys():\n",
    "\t#print(f'Calculating p-value for group: {group}')\n",
    "\tassert len(all_cpgs[group]['angus_ref_methylation']) == len(all_cpgs[group]['brahman_ref_methylation']), \"Downsampling failed\"\n",
    "\t_, pvalue = stats.wilcoxon(all_cpgs[group]['angus_ref_methylation'], all_cpgs[group]['brahman_ref_methylation'])\n",
    "\t#print(f\"Group: {group}, p-value: {pvalue}\")\n",
    "\tgroup_p_vals['group'].append(group)\n",
    "\tgroup_p_vals['n_CpGs'].append(len(all_cpgs[group]['angus_ref_methylation']))\n",
    "\tgroup_p_vals['angus_methylation_mean'].append(np.mean(all_cpgs[group]['angus_ref_methylation']))\n",
    "\tgroup_p_vals['brahman_methylation_mean'].append(np.mean(all_cpgs[group]['brahman_ref_methylation']))\n",
    "\tgroup_p_vals['angus_methylation_std_dev'].append(np.std(all_cpgs[group]['angus_ref_methylation']))\n",
    "\tgroup_p_vals['brahman_methylation_std_dev'].append(np.std(all_cpgs[group]['brahman_ref_methylation']))\n",
    "\tgroup_p_vals['difference'].append(np.abs(np.mean(all_cpgs[group]['angus_ref_methylation']) - np.mean(all_cpgs[group]['brahman_ref_methylation'])))\n",
    "\tgroup_p_vals['p_value'].append(pvalue)\n",
    "\tpvals.append(pvalue)\n",
    "\n",
    "# Test the p-values for multiple testing\n",
    "multipletests(pvals, alpha=0.01, method='fdr_bh')\n",
    "group_p_vals['adjusted_p_value'] = multipletests(pvals, alpha=0.01, method='fdr_bh')[1]\n",
    "pd.DataFrame(group_p_vals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 24/24 [00:16<00:00,  1.46it/s]\n",
      "100%|██████████| 24/24 [00:16<00:00,  1.45it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "# descriptive_stats_dict = {}\n",
    "# means = {'sampleID':[],'methylation':[],'reference':[], 'group':[]}\n",
    "# for sample in tqdm(angus_all.columns[:-1]):\n",
    "# \t#methylation_dict = {'sampleID':[],'methylation':[],'reference':[],\n",
    "# \t#'group':[]}\n",
    "\t\n",
    "# \t# get the methylation\n",
    "# \tmethylation = angus_all[sample].values\n",
    "# \tmeth_stats = angus_all[sample].describe()\n",
    "# \t# get the reference\n",
    "# \treference = 'Angus'\n",
    "# \t# get the sampleID\n",
    "# \tsampleID = sample.split(\"_\")[1]\n",
    "# \tgroup = sample.split('_')[0]\n",
    "# \t# append to the dictionary\n",
    "# \t# methylation_dict['sampleID'].extend(sampleID)\n",
    "# \t# methylation_dict['methylation'].extend(list(methylation))\n",
    "# \t# methylation_dict['reference'].extend(list(reference))\n",
    "# \t# methylation_dict['group'].extend(list(group))\n",
    "# \tdescriptive_stats_dict[f'{sampleID}_{reference}'] = meth_stats\n",
    "# \tmeans['sampleID'].append(sampleID)\n",
    "# \tmeans['methylation'].append(np.mean(methylation))\n",
    "# \tmeans['reference'].append(reference)\n",
    "# \tmeans['group'].append(group)\n",
    "# \t# pd.DataFrame.from_dict(methylation_dict).to_csv(\n",
    "# \t# \tplotting_file,\n",
    "# \t# \theader=False,\n",
    "# \t# \tindex=False)\n",
    "\n",
    "# del angus_all\n",
    "# gc.collect()\n",
    "\n",
    "# for sample in tqdm(brahman_all.columns[:-1]):\n",
    "# \t#methylation_dict = {'sampleID':[],'methylation':[],'reference':[],\n",
    "# \t#'group':[]}\n",
    "# \t# get the methylation\n",
    "# \tmethylation = brahman_all[sample].values\n",
    "# \tmeth_stats = brahman_all[sample].describe()\n",
    "# \t# get the reference\n",
    "# \treference = 'Brahman'\n",
    "# \t# get the sampleID\n",
    "# \tsampleID = sample.split(\"_\")[1]\n",
    "# \tgroup = sample.split('_')[0]\n",
    "# \t# append to the dictionary\n",
    "# \t# methylation_dict['sampleID'].extend(sampleID)\n",
    "# \t# methylation_dict['methylation'].extend(list(methylation))\n",
    "# \t# methylation_dict['reference'].extend(list(reference))\n",
    "# \t# methylation_dict['group'].extend(list(group))\n",
    "# \tdescriptive_stats_dict[f'{sampleID}_{reference}'] = meth_stats\n",
    "# \tmeans['sampleID'].append(sampleID)\n",
    "# \tmeans['methylation'].append(np.mean(methylation))\n",
    "# \tmeans['reference'].append(reference)\n",
    "# \tmeans['group'].append(group)\n",
    "# \t# pd.DataFrame.from_dict(methylation_dict).to_csv(\n",
    "# \t# \tplotting_file,\n",
    "# \t# \theader=False,\n",
    "# \t# \tindex=False)\n",
    "\n",
    "# del brahman_all\n",
    "# gc.collect()\n",
    "\n",
    "# print('Done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame.from_dict(means)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colour_maps = ['dark','muted','bright','pastel']\n",
    "for n, group in enumerate(['AxA', 'AxB','BxA','BxB']):\n",
    "\tplt.close()\n",
    "\tsns.set_style('whitegrid')\n",
    "\n",
    "\tmosaic_str = '''\n",
    "\tA\n",
    "\tB\n",
    "\t'''\n",
    "\n",
    "\tfig = plt.figure(constrained_layout=True)\n",
    "\tfig, ax = plt.subplot_mosaic(mosaic_str, \n",
    "\t\t\t\t\t\t\t gridspec_kw={'height_ratios': [3, 0.1]})\n",
    "\n",
    "\tsns.swarmplot(data=df[df['group'] == group],\n",
    "\t\t\tx='reference', y='methylation_mean',\n",
    "\t\t\thue='sampleID', ax=ax['A'],\n",
    "\t\t\tpalette=colour_maps[n])\n",
    "\tsns.swarmplot(data=df[df['group'] == group],\n",
    "\t\t\tx='reference', y='methylation_mean',\n",
    "\t\t\thue='sampleID', ax=ax['B'],\n",
    "\t\t\tpalette=colour_maps[n])\n",
    "\tfig.set_size_inches(2,4)\n",
    "\tax['A'].set_ylim(45, 70)\n",
    "\tax['B'].set_ylim(0,20)\n",
    "\tax['A'].get_xaxis().set_visible(False)\n",
    "\tax['A'].set_ylabel(\"Methylation %\")\n",
    "\tax['B'].set_ylabel(\"\")\n",
    "\tax['B'].set_xlabel(\"\")\n",
    "\tax['A'].legend(loc=(1.01,0.48), title='Sample ID')\n",
    "\tax['A'].xaxis.tick_top()\n",
    "\tax['B'].xaxis.tick_bottom()\n",
    "\tax['B'].get_legend().remove()\n",
    "\tax['A'].set_title(f\"Group = {group}\")\n",
    "\tax['B'].set_yticks([])\n",
    "\tax['B'].set_xlabel(\"Reference\")\n",
    "\tplt.savefig(f'/Users/callummacphillamy/Library/CloudStorage/OneDrive-UniversityofAdelaide/PhD/methylation_chapter/GigaScience_submission/figures/{group}_allCpGs_methylation.png', dpi=400, bbox_inches='tight')\n",
    "\tplt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>methylation_mean</th>\n",
       "      <th>methylation_median</th>\n",
       "      <th>reference</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>F22</td>\n",
       "      <td>51.339325</td>\n",
       "      <td>56.4516</td>\n",
       "      <td>Angus</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>F22</td>\n",
       "      <td>51.677111</td>\n",
       "      <td>56.6667</td>\n",
       "      <td>Brahman</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>F46</td>\n",
       "      <td>60.019766</td>\n",
       "      <td>68.0851</td>\n",
       "      <td>Angus</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>F46</td>\n",
       "      <td>60.353790</td>\n",
       "      <td>68.2927</td>\n",
       "      <td>Brahman</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>F56</td>\n",
       "      <td>49.711692</td>\n",
       "      <td>54.2857</td>\n",
       "      <td>Angus</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>F56</td>\n",
       "      <td>50.005140</td>\n",
       "      <td>54.5455</td>\n",
       "      <td>Brahman</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>F65</td>\n",
       "      <td>49.966933</td>\n",
       "      <td>54.8387</td>\n",
       "      <td>Angus</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>F65</td>\n",
       "      <td>50.364694</td>\n",
       "      <td>55.0000</td>\n",
       "      <td>Brahman</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>F78</td>\n",
       "      <td>52.664885</td>\n",
       "      <td>58.3333</td>\n",
       "      <td>Angus</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>F78</td>\n",
       "      <td>53.029043</td>\n",
       "      <td>58.6207</td>\n",
       "      <td>Brahman</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>F99</td>\n",
       "      <td>52.658632</td>\n",
       "      <td>58.3333</td>\n",
       "      <td>Angus</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>F99</td>\n",
       "      <td>52.994536</td>\n",
       "      <td>58.5366</td>\n",
       "      <td>Brahman</td>\n",
       "      <td>BxB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sampleID  methylation_mean  methylation_median reference group\n",
       "36      F22         51.339325             56.4516     Angus   BxB\n",
       "37      F22         51.677111             56.6667   Brahman   BxB\n",
       "38      F46         60.019766             68.0851     Angus   BxB\n",
       "39      F46         60.353790             68.2927   Brahman   BxB\n",
       "40      F56         49.711692             54.2857     Angus   BxB\n",
       "41      F56         50.005140             54.5455   Brahman   BxB\n",
       "42      F65         49.966933             54.8387     Angus   BxB\n",
       "43      F65         50.364694             55.0000   Brahman   BxB\n",
       "44      F78         52.664885             58.3333     Angus   BxB\n",
       "45      F78         53.029043             58.6207   Brahman   BxB\n",
       "46      F99         52.658632             58.3333     Angus   BxB\n",
       "47      F99         52.994536             58.5366   Brahman   BxB"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group: AxA\n",
      "Test Statistic: 0.0\n",
      "P-value: 0.03125\n",
      "\n",
      "Group: AxB\n",
      "Test Statistic: 1.0\n",
      "P-value: 0.0625\n",
      "\n",
      "Group: BxA\n",
      "Test Statistic: 0.0\n",
      "P-value: 0.03125\n",
      "\n",
      "Group: BxB\n",
      "Test Statistic: 0.0\n",
      "P-value: 0.03125\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for group in ['AxA', 'AxB','BxA','BxB']:\n",
    "\tgroup_df = df[df['group'] == group]\n",
    "\tangus = group_df[group_df['reference'] == 'Angus']\n",
    "\tbrahman = group_df[group_df['reference'] == 'Brahman']\n",
    "\tangus_meth = angus['methylation_mean'].values\n",
    "\tbrahman_meth = brahman['methylation_mean'].values\n",
    "\ttest, p = stats.wilcoxon(angus_meth, brahman_meth)\n",
    "\tprint('Group:', group)\n",
    "\tprint('Test Statistic:', test)\n",
    "\tprint('P-value:', p)\n",
    "\tprint('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shared CpGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the data representing methylation when aligned to Angus\n",
    "angus_all = filterParquet('./aligned2Angus/All.samples.Consensus.meth.10X.parquet')\n",
    "brahman_all = filterParquet('./aligned2Brahman/All.samples.Consensus.meth.10X.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample: AxA_F103\n",
      "Quantification bias: 0.0564830957259872\n",
      "Sample: AxA_F105\n",
      "Quantification bias: 0.06455779129370853\n",
      "Sample: AxA_F52\n",
      "Quantification bias: 0.06274313680675192\n",
      "Sample: AxA_F53\n",
      "Quantification bias: 0.049638179956417616\n",
      "Sample: AxA_F60\n",
      "Quantification bias: 0.044076632715267294\n",
      "Sample: AxA_F7\n",
      "Quantification bias: 0.05164322907960184\n",
      "Sample: AxB_F100\n",
      "Quantification bias: 0.04284799230181591\n",
      "Sample: AxB_F104\n",
      "Quantification bias: 0.03288743230156399\n",
      "Sample: AxB_F106\n",
      "Quantification bias: 0.03582452031492522\n",
      "Sample: AxB_F61\n",
      "Quantification bias: 0.03815921084444085\n",
      "Sample: AxB_F74\n",
      "Quantification bias: 0.038663329894014\n",
      "Sample: AxB_F97\n",
      "Quantification bias: 0.025899768595458283\n",
      "Sample: BxA_F13\n",
      "Quantification bias: 0.029549184667231998\n",
      "Sample: BxA_F62\n",
      "Quantification bias: 0.028313696589876258\n",
      "Sample: BxA_F77\n",
      "Quantification bias: 0.03109646924883691\n",
      "Sample: BxA_F80\n",
      "Quantification bias: 0.03514137468794587\n",
      "Sample: BxA_F8\n",
      "Quantification bias: 0.03825622704746934\n",
      "Sample: BxA_F91\n",
      "Quantification bias: 0.027611977230414197\n",
      "Sample: BxB_F22\n",
      "Quantification bias: 0.016386727667651257\n",
      "Sample: BxB_F46\n",
      "Quantification bias: 0.029884010617486467\n",
      "Sample: BxB_F56\n",
      "Quantification bias: 0.017034227386098166\n",
      "Sample: BxB_F65\n",
      "Quantification bias: 0.015951046169483618\n",
      "Sample: BxB_F78\n",
      "Quantification bias: 0.023566277935920255\n",
      "Sample: BxB_F99\n",
      "Quantification bias: 0.021221850765478223\n"
     ]
    }
   ],
   "source": [
    "# Get the methylation of every site in the parquet file and\n",
    "# for each sample and compare, at the sample level if there is a\n",
    "# significant difference in the methylation distribution\n",
    "\n",
    "# Get the methylation of every site in the parquet file and\n",
    "# for each sample and compare, at the sample level if there is a\n",
    "# significant difference in the methylation distribution\n",
    "\n",
    "p_values = {sample: [] for sample in angus_all.columns[:-1]}\n",
    "\n",
    "means = {'sampleID':[],'group':[],\n",
    "\t\t 'n_CpGs':[],\n",
    "\t\t 'angus_methylation_mean':[],\n",
    "\t\t 'brahman_methylation_mean':[],\n",
    "\t\t 'angus_methylation_std':[],\n",
    "\t\t 'brahman_methylation_std':[],\n",
    "\t\t 'difference':[],\n",
    "\t\t 'p_value':[],\n",
    "\t\t 'adjusted_p_value':[]}\n",
    "\n",
    "\n",
    "\n",
    "for sample in angus_all.columns[:-1]:\n",
    "\n",
    "\tgroup = sample.split(\"_\")[0]\n",
    "\tsampleID = sample.split(\"_\")[1]\n",
    "\n",
    "\t# Create an index of each genome for any necessary\n",
    "\t# downsampling\n",
    "\tangus_all_idx = np.arange(angus_all.shape[0])\n",
    "\tbrahman_all_idx = np.arange(brahman_all.shape[0])\n",
    "\tif len(angus_all_idx) > len(brahman_all_idx):\n",
    "\t\tangus_all_idx = np.random.choice(angus_all_idx, len(brahman_all_idx), replace=False)\n",
    "\telif len(angus_all_idx) < len(brahman_all_idx):\n",
    "\t\tbrahman_all_idx = np.random.choice(brahman_all_idx, len(angus_all_idx), replace=False)\n",
    "\n",
    "\tmapped_2_angus_methylation = angus_all[sample].values\n",
    "\tmapped_2_brahman_methylation = brahman_all[sample].values\n",
    "\n",
    "\t# Downsample the methylation arrays\n",
    "\tmapped_2_angus_methylation = mapped_2_angus_methylation[angus_all_idx]\n",
    "\tmapped_2_brahman_methylation = mapped_2_brahman_methylation[brahman_all_idx]\n",
    "\n",
    "\t# Compute the mean and median methylation for each sample\n",
    "\t# and genome\n",
    "\tmeans['sampleID'].append(sampleID)\n",
    "\tmeans['n_CpGs'].append(len(mapped_2_angus_methylation))\n",
    "\tmeans['angus_methylation_mean'].append(mapped_2_angus_methylation.mean())\n",
    "\tmeans['angus_methylation_std'].append(np.std(mapped_2_angus_methylation))\n",
    "\t#means['reference'].append('Angus')\n",
    "\tmeans['group'].append(group)\n",
    "\t\n",
    "\t#means['sampleID'].append(sampleID)\n",
    "\tmeans['brahman_methylation_mean'].append(mapped_2_brahman_methylation.mean())\n",
    "\tmeans['brahman_methylation_std'].append(np.std(mapped_2_brahman_methylation))\n",
    "\t#means['reference'].append('Brahman')\n",
    "\t#means['group'].append(group)\n",
    "\n",
    "\n",
    "\t# Compare the distributions of each sample\n",
    "\t_, pvalue = stats.wilcoxon(mapped_2_angus_methylation, mapped_2_brahman_methylation)\n",
    "\tp_values[sample].append(pvalue)\n",
    "\tmeans['p_value'].append(pvalue)\n",
    "\n",
    "\t# Get the percentage difference between the means\n",
    "\t# of the two distributions\n",
    "\t#percentage_diff = ((mapped_2_angus_methylation.mean() - mapped_2_brahman_methylation.mean()) / (mapped_2_angus_methylation.mean() + mapped_2_brahman_methylation.mean()/2)) * 100\n",
    "\t#difference = np.mean((np.abs(mapped_2_angus_methylation - mapped_2_brahman_methylation)))\n",
    "\tprint(f\"Sample: {sample}\\nQuantification bias: {np.abs(mapped_2_angus_methylation.mean() - mapped_2_brahman_methylation.mean())}\")\n",
    "\tdifference = np.abs(mapped_2_angus_methylation.mean() - mapped_2_brahman_methylation.mean())\n",
    "\t#means['difference'].append(np.abs(percentage_diff))\n",
    "\tmeans['difference'].append(difference)\n",
    "\t#print(f\"Sample: {sample}, p-value: {pvalue}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>group</th>\n",
       "      <th>n_CpGs</th>\n",
       "      <th>angus_methylation_mean</th>\n",
       "      <th>brahman_methylation_mean</th>\n",
       "      <th>angus_methylation_std</th>\n",
       "      <th>brahman_methylation_std</th>\n",
       "      <th>difference</th>\n",
       "      <th>p_value</th>\n",
       "      <th>adjusted_p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>F103</td>\n",
       "      <td>AxA</td>\n",
       "      <td>16204834</td>\n",
       "      <td>57.057316</td>\n",
       "      <td>57.000833</td>\n",
       "      <td>24.437498</td>\n",
       "      <td>24.465181</td>\n",
       "      <td>0.056483</td>\n",
       "      <td>1.235371e-09</td>\n",
       "      <td>7.412226e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>F105</td>\n",
       "      <td>AxA</td>\n",
       "      <td>16204834</td>\n",
       "      <td>62.081784</td>\n",
       "      <td>62.017226</td>\n",
       "      <td>25.810093</td>\n",
       "      <td>25.843298</td>\n",
       "      <td>0.064558</td>\n",
       "      <td>3.776387e-13</td>\n",
       "      <td>9.063328e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>F52</td>\n",
       "      <td>AxA</td>\n",
       "      <td>16204834</td>\n",
       "      <td>57.873408</td>\n",
       "      <td>57.810665</td>\n",
       "      <td>24.353167</td>\n",
       "      <td>24.381333</td>\n",
       "      <td>0.062743</td>\n",
       "      <td>2.122180e-12</td>\n",
       "      <td>2.546616e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>F53</td>\n",
       "      <td>AxA</td>\n",
       "      <td>16204834</td>\n",
       "      <td>53.734503</td>\n",
       "      <td>53.684865</td>\n",
       "      <td>23.452634</td>\n",
       "      <td>23.481419</td>\n",
       "      <td>0.049638</td>\n",
       "      <td>1.363609e-08</td>\n",
       "      <td>6.545324e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>F60</td>\n",
       "      <td>AxA</td>\n",
       "      <td>16204834</td>\n",
       "      <td>50.491531</td>\n",
       "      <td>50.447455</td>\n",
       "      <td>23.002033</td>\n",
       "      <td>23.028587</td>\n",
       "      <td>0.044077</td>\n",
       "      <td>4.053988e-08</td>\n",
       "      <td>1.621595e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sampleID group    n_CpGs  angus_methylation_mean  brahman_methylation_mean  \\\n",
       "0     F103   AxA  16204834               57.057316                 57.000833   \n",
       "1     F105   AxA  16204834               62.081784                 62.017226   \n",
       "2      F52   AxA  16204834               57.873408                 57.810665   \n",
       "3      F53   AxA  16204834               53.734503                 53.684865   \n",
       "4      F60   AxA  16204834               50.491531                 50.447455   \n",
       "\n",
       "   angus_methylation_std  brahman_methylation_std  difference       p_value  \\\n",
       "0              24.437498                24.465181    0.056483  1.235371e-09   \n",
       "1              25.810093                25.843298    0.064558  3.776387e-13   \n",
       "2              24.353167                24.381333    0.062743  2.122180e-12   \n",
       "3              23.452634                23.481419    0.049638  1.363609e-08   \n",
       "4              23.002033                23.028587    0.044077  4.053988e-08   \n",
       "\n",
       "   adjusted_p_value  \n",
       "0      7.412226e-09  \n",
       "1      9.063328e-12  \n",
       "2      2.546616e-11  \n",
       "3      6.545324e-08  \n",
       "4      1.621595e-07  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test the p-values for multiple testing\n",
    "multipletests(pd.DataFrame(p_values).values.flatten(), alpha=0.01, method='fdr_bh')\n",
    "means['adjusted_p_value'].extend(multipletests(pd.DataFrame(p_values).values.flatten(), alpha=0.01, method='fdr_bh')[1])\n",
    "pd.DataFrame(means).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(means).to_csv('/Users/callummacphillamy/Library/CloudStorage/OneDrive-UniversityofAdelaide/PhD/methylation_chapter/GigaScience_submission/supplementary_tables/shared_CpGs_sample-wise.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/24 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 24/24 [03:39<00:00,  9.14s/it]\n"
     ]
    }
   ],
   "source": [
    "# S table re. Figure3A variance and what not\n",
    "all_cpgs = {'AxA':{'angus_ref_methylation':[],\n",
    "\t\t\t\t   'brahman_ref_methylation':[]},\n",
    "\t\t\t'AxB':{'angus_ref_methylation':[],\n",
    "\t\t\t\t   'brahman_ref_methylation':[]},\n",
    "\t\t\t'BxA':{'angus_ref_methylation':[],\n",
    "\t\t           'brahman_ref_methylation':[]},\n",
    "\t\t\t'BxB':{'angus_ref_methylation':[],\n",
    "\t\t           'brahman_ref_methylation':[]}}\n",
    "\n",
    "stable_dict = {'sampleID':[],\n",
    "\t\t\t   'n_CpGs':[],\n",
    "\t\t\t   'angus_methylation_mean':[],\n",
    "\t\t\t   'brahman_methylation_mean':[],\n",
    "\t\t\t   'angus_methylation_std_dev':[],\n",
    "\t\t\t   'brahman_methylation_std_dev':[],\n",
    "\t\t\t   'difference':[],\n",
    "\t\t\t   'p_value':[],\n",
    "\t\t\t   'group':[]}\n",
    "for sample in tqdm(angus_all.columns[:-1]):\n",
    "\n",
    "\tgroup = sample.split(\"_\")[0]\n",
    "\tsampleID = sample.split(\"_\")[1]\n",
    "\n",
    "\t# Create an index of each genome for any necessary\n",
    "\t# downsampling\n",
    "\tangus_all_idx = np.arange(angus_all.shape[0])\n",
    "\tbrahman_all_idx = np.arange(brahman_all.shape[0])\n",
    "\tif len(angus_all_idx) > len(brahman_all_idx):\n",
    "\t\tangus_all_idx = np.random.choice(angus_all_idx, len(brahman_all_idx), replace=False)\n",
    "\telif len(angus_all_idx) < len(brahman_all_idx):\n",
    "\t\tbrahman_all_idx = np.random.choice(brahman_all_idx, len(angus_all_idx), replace=False)\n",
    "\n",
    "\tmapped_2_angus_methylation = angus_all[sample].values\n",
    "\tmapped_2_brahman_methylation = brahman_all[sample].values\n",
    "\n",
    "\t# Downsample the methylation arrays\n",
    "\tmapped_2_angus_methylation = mapped_2_angus_methylation[angus_all_idx]\n",
    "\tmapped_2_brahman_methylation = mapped_2_brahman_methylation[brahman_all_idx]\n",
    "\n",
    "\t# Compute the mean and median methylation for each sample\n",
    "\t# and genome\n",
    "\tstable_dict['sampleID'].append(sampleID)\n",
    "\tassert len(mapped_2_angus_methylation) == len(mapped_2_brahman_methylation), \"Downsampling failed\"\n",
    "\tstable_dict['n_CpGs'].append(len(mapped_2_angus_methylation))\n",
    "\tstable_dict['angus_methylation_mean'].append(mapped_2_angus_methylation.mean())\n",
    "\tstable_dict['brahman_methylation_mean'].append(mapped_2_brahman_methylation.mean())\n",
    "\tstable_dict['angus_methylation_std_dev'].append(np.std(mapped_2_angus_methylation))\n",
    "\tstable_dict['brahman_methylation_std_dev'].append(np.std(mapped_2_brahman_methylation))\n",
    "\tstable_dict['difference'].append(np.abs(mapped_2_angus_methylation.mean() - mapped_2_brahman_methylation.mean()))\n",
    "\tstable_dict['p_value'].append(stats.wilcoxon(mapped_2_angus_methylation, mapped_2_brahman_methylation).pvalue)\n",
    "\tstable_dict['group'].append(group)\n",
    "\t\n",
    "\t# stable_dict['sampleID'].append(sampleID)\n",
    "\t# stable_dict['methylation_mean'].append(mapped_2_brahman_methylation.mean())\n",
    "\t# stable_dict['methylation_std_dev'].append(np.std(mapped_2_brahman_methylation))\n",
    "\t# stable_dict['reference'].append('Brahman')\n",
    "\t# stable_dict['group'].append(group)\n",
    "\n",
    "\tall_cpgs[group]['angus_ref_methylation'].extend(mapped_2_angus_methylation)\n",
    "\tall_cpgs[group]['brahman_ref_methylation'].extend(mapped_2_brahman_methylation)\n",
    "#pd.DataFrame.from_dict(stable_dict).to_csv('/Users/callummacphillamy/Library/CloudStorage/OneDrive-UniversityofAdelaide/PhD/methylation_chapter/GigaScience_submission/supplementary_tables/shared_CpGs_per_sample.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>n_CpGs</th>\n",
       "      <th>angus_methylation_mean</th>\n",
       "      <th>brahman_methylation_mean</th>\n",
       "      <th>angus_methylation_std_dev</th>\n",
       "      <th>brahman_methylation_std_dev</th>\n",
       "      <th>difference</th>\n",
       "      <th>p_value</th>\n",
       "      <th>adjusted_p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AxA</td>\n",
       "      <td>97229004</td>\n",
       "      <td>55.872572</td>\n",
       "      <td>55.817859</td>\n",
       "      <td>24.405071</td>\n",
       "      <td>24.432431</td>\n",
       "      <td>0.054713</td>\n",
       "      <td>3.540816e-54</td>\n",
       "      <td>1.416326e-53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AxB</td>\n",
       "      <td>97229004</td>\n",
       "      <td>52.909838</td>\n",
       "      <td>52.873993</td>\n",
       "      <td>23.446976</td>\n",
       "      <td>23.478456</td>\n",
       "      <td>0.035845</td>\n",
       "      <td>3.118312e-22</td>\n",
       "      <td>6.236625e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BxA</td>\n",
       "      <td>97229004</td>\n",
       "      <td>51.141253</td>\n",
       "      <td>51.109286</td>\n",
       "      <td>23.279071</td>\n",
       "      <td>23.308486</td>\n",
       "      <td>0.031966</td>\n",
       "      <td>2.471715e-17</td>\n",
       "      <td>3.295620e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BxB</td>\n",
       "      <td>97229004</td>\n",
       "      <td>53.942699</td>\n",
       "      <td>53.921943</td>\n",
       "      <td>24.545691</td>\n",
       "      <td>24.580179</td>\n",
       "      <td>0.020756</td>\n",
       "      <td>5.794265e-05</td>\n",
       "      <td>5.794265e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  group    n_CpGs  angus_methylation_mean  brahman_methylation_mean  \\\n",
       "0   AxA  97229004               55.872572                 55.817859   \n",
       "1   AxB  97229004               52.909838                 52.873993   \n",
       "2   BxA  97229004               51.141253                 51.109286   \n",
       "3   BxB  97229004               53.942699                 53.921943   \n",
       "\n",
       "   angus_methylation_std_dev  brahman_methylation_std_dev  difference  \\\n",
       "0                  24.405071                    24.432431    0.054713   \n",
       "1                  23.446976                    23.478456    0.035845   \n",
       "2                  23.279071                    23.308486    0.031966   \n",
       "3                  24.545691                    24.580179    0.020756   \n",
       "\n",
       "        p_value  adjusted_p_value  \n",
       "0  3.540816e-54      1.416326e-53  \n",
       "1  3.118312e-22      6.236625e-22  \n",
       "2  2.471715e-17      3.295620e-17  \n",
       "3  5.794265e-05      5.794265e-05  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_p_vals = {'group':[],\n",
    "\t\t\t\t'n_CpGs':[],\n",
    "\t\t\t\t'angus_methylation_mean':[],\n",
    "\t\t\t\t'brahman_methylation_mean':[],\n",
    "\t\t\t\t'angus_methylation_std_dev':[],\n",
    "\t\t\t\t'brahman_methylation_std_dev':[],\n",
    "\t\t\t\t'difference':[],\n",
    "\t\t\t\t'p_value':[],\n",
    "\t\t\t\t'adjusted_p_value':[]}\n",
    "\n",
    "p_values = []\n",
    "\n",
    "for group in all_cpgs.keys():\n",
    "\t#print(f'Calculating p-value for group: {group}')\n",
    "\tassert len(all_cpgs[group]['angus_ref_methylation']) == len(all_cpgs[group]['brahman_ref_methylation']), \"Downsampling failed\"\n",
    "\t_, pvalue = stats.wilcoxon(all_cpgs[group]['angus_ref_methylation'], all_cpgs[group]['brahman_ref_methylation'])\n",
    "\t#print(f\"Group: {group}, p-value: {pvalue}\")\n",
    "\tgroup_p_vals['group'].append(group)\n",
    "\tgroup_p_vals['n_CpGs'].append(len(all_cpgs[group]['angus_ref_methylation']))\n",
    "\tgroup_p_vals['angus_methylation_mean'].append(np.mean(all_cpgs[group]['angus_ref_methylation']))\n",
    "\tgroup_p_vals['brahman_methylation_mean'].append(np.mean(all_cpgs[group]['brahman_ref_methylation']))\n",
    "\tgroup_p_vals['angus_methylation_std_dev'].append(np.std(all_cpgs[group]['angus_ref_methylation']))\n",
    "\tgroup_p_vals['brahman_methylation_std_dev'].append(np.std(all_cpgs[group]['brahman_ref_methylation']))\n",
    "\tgroup_p_vals['difference'].append(np.abs(np.mean(all_cpgs[group]['angus_ref_methylation']) - np.mean(all_cpgs[group]['brahman_ref_methylation'])))\n",
    "\tgroup_p_vals['p_value'].append(pvalue)\n",
    "\tp_values.append(pvalue)\n",
    "group_p_vals['adjusted_p_value'].extend(multipletests(p_values, alpha=0.01, method='fdr_bh')[1])\n",
    "pd.DataFrame(group_p_vals).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame.from_dict(group_p_vals).to_csv('/Users/callummacphillamy/Library/CloudStorage/OneDrive-UniversityofAdelaide/PhD/methylation_chapter/GigaScience_submission/supplementary_tables/shared_CpGs_per_group.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = pd.DataFrame.from_dict(means)\n",
    "\n",
    "colour_maps = ['dark','muted','bright','pastel']\n",
    "for n, group in enumerate(['AxA', 'AxB','BxA','BxB']):\n",
    "\tplt.close()\n",
    "\tsns.set_style('whitegrid')\n",
    "\n",
    "\tmosaic_str = '''\n",
    "\tA\n",
    "\tB\n",
    "\t'''\n",
    "\n",
    "\tfig = plt.figure(constrained_layout=True)\n",
    "\tfig, ax = plt.subplot_mosaic(mosaic_str, \n",
    "\t\t\t\t\t\t\t gridspec_kw={'height_ratios': [3, 0.1]})\n",
    "\n",
    "\tsns.swarmplot(data=df[df['group'] == group],\n",
    "\t\t\tx='reference', y='methylation_mean',\n",
    "\t\t\thue='sampleID', ax=ax['A'],\n",
    "\t\t\tpalette=colour_maps[n])\n",
    "\tsns.swarmplot(data=df[df['group'] == group],\n",
    "\t\t\tx='reference', y='methylation_mean',\n",
    "\t\t\thue='sampleID', ax=ax['B'],\n",
    "\t\t\tpalette=colour_maps[n])\n",
    "\tfig.set_size_inches(2,4)\n",
    "\tax['A'].set_ylim(45, 70)\n",
    "\tax['B'].set_ylim(0,20)\n",
    "\tax['A'].get_xaxis().set_visible(False)\n",
    "\tax['A'].set_ylabel(\"Methylation %\")\n",
    "\tax['B'].set_ylabel(\"\")\n",
    "\tax['B'].set_xlabel(\"\")\n",
    "\tax['A'].legend(loc=(1.01,0.48), title='Sample ID')\n",
    "\tax['A'].xaxis.tick_top()\n",
    "\tax['B'].xaxis.tick_bottom()\n",
    "\tax['B'].get_legend().remove()\n",
    "\tax['A'].set_title(f\"Group = {group}\")\n",
    "\tax['B'].set_yticks([])\n",
    "\tax['B'].set_xlabel(\"Reference\")\n",
    "\tplt.savefig(f'/Users/callummacphillamy/Library/CloudStorage/OneDrive-UniversityofAdelaide/PhD/methylation_chapter/GigaScience_submission/figures/{group}_SharedCpGs_methylation.png', dpi=400, bbox_inches='tight')\n",
    "\tplt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group: AxA\n",
      "Test Statistic: 0.0\n",
      "P-value: 0.31731050786291415\n",
      "\n",
      "Group: AxB\n",
      "Test Statistic: 0.0\n",
      "P-value: 0.31731050786291415\n",
      "\n",
      "Group: BxA\n",
      "Test Statistic: 0.0\n",
      "P-value: 0.17971249487899976\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/callummacphillamy/mambaforge/envs/wgbs/lib/python3.10/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: Exact p-value calculation does not work if there are zeros. Switching to normal approximation.\n",
      "  res = hypotest_fun_out(*samples, **kwds)\n",
      "/Users/callummacphillamy/mambaforge/envs/wgbs/lib/python3.10/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: Sample size too small for normal approximation.\n",
      "  res = hypotest_fun_out(*samples, **kwds)\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "zero_method 'wilcox' and 'pratt' do not work if x - y is zero for all elements.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 7\u001b[0m\n\u001b[1;32m      5\u001b[0m angus_meth \u001b[38;5;241m=\u001b[39m angus[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmethylation_median\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mvalues\n\u001b[1;32m      6\u001b[0m brahman_meth \u001b[38;5;241m=\u001b[39m brahman[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmethylation_median\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mvalues\n\u001b[0;32m----> 7\u001b[0m test, p \u001b[38;5;241m=\u001b[39m \u001b[43mstats\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwilcoxon\u001b[49m\u001b[43m(\u001b[49m\u001b[43mangus_meth\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbrahman_meth\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGroup:\u001b[39m\u001b[38;5;124m'\u001b[39m, group)\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTest Statistic:\u001b[39m\u001b[38;5;124m'\u001b[39m, test)\n",
      "File \u001b[0;32m~/mambaforge/envs/wgbs/lib/python3.10/site-packages/scipy/_lib/_util.py:779\u001b[0m, in \u001b[0;36m_rename_parameter.<locals>.decorator.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    777\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(message)\n\u001b[1;32m    778\u001b[0m     kwargs[new_name] \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mpop(old_name)\n\u001b[0;32m--> 779\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfun\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/mambaforge/envs/wgbs/lib/python3.10/site-packages/scipy/stats/_axis_nan_policy.py:531\u001b[0m, in \u001b[0;36m_axis_nan_policy_factory.<locals>.axis_nan_policy_decorator.<locals>.axis_nan_policy_wrapper\u001b[0;34m(***failed resolving arguments***)\u001b[0m\n\u001b[1;32m    529\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m sentinel:\n\u001b[1;32m    530\u001b[0m     samples \u001b[38;5;241m=\u001b[39m _remove_sentinel(samples, paired, sentinel)\n\u001b[0;32m--> 531\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43mhypotest_fun_out\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msamples\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    532\u001b[0m res \u001b[38;5;241m=\u001b[39m result_to_tuple(res)\n\u001b[1;32m    533\u001b[0m res \u001b[38;5;241m=\u001b[39m _add_reduced_axes(res, reduced_axes, keepdims)\n",
      "File \u001b[0;32m~/mambaforge/envs/wgbs/lib/python3.10/site-packages/scipy/stats/_morestats.py:4118\u001b[0m, in \u001b[0;36mwilcoxon\u001b[0;34m(x, y, zero_method, correction, alternative, method)\u001b[0m\n\u001b[1;32m   4116\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m zero_method \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwilcox\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpratt\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m   4117\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m n_zero \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mlen\u001b[39m(d):\n\u001b[0;32m-> 4118\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mzero_method \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwilcox\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m and \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpratt\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m do not \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   4119\u001b[0m                          \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwork if x - y is zero for all elements.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   4120\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m zero_method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwilcox\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m   4121\u001b[0m     \u001b[38;5;66;03m# Keep all non-zero differences\u001b[39;00m\n\u001b[1;32m   4122\u001b[0m     d \u001b[38;5;241m=\u001b[39m compress(np\u001b[38;5;241m.\u001b[39mnot_equal(d, \u001b[38;5;241m0\u001b[39m), d)\n",
      "\u001b[0;31mValueError\u001b[0m: zero_method 'wilcox' and 'pratt' do not work if x - y is zero for all elements."
     ]
    }
   ],
   "source": [
    "for group in ['AxA', 'AxB','BxA','BxB']:\n",
    "\tgroup_df = df[df['group'] == group]\n",
    "\tangus = group_df[group_df['reference'] == 'Angus']\n",
    "\tbrahman = group_df[group_df['reference'] == 'Brahman']\n",
    "\tangus_meth = angus['methylation_mean'].values\n",
    "\tbrahman_meth = brahman['methylation_mean'].values\n",
    "\ttest, p = stats.wilcoxon(angus_meth, brahman_meth)\n",
    "\tprint('Group:', group)\n",
    "\tprint('Test Statistic:', test)\n",
    "\tprint('P-value:', p)\n",
    "\tprint('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wgbs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
