{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "import numpy as np\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MethPipe input\n",
    "col 1: chromosome  \n",
    "col 2: position of Cytosine  \n",
    "col 3: strand  \n",
    "col 4: sequence context  \n",
    "col 5: proportion of reads reporting C at that site  \n",
    "col 6: coverage  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "methylkit_files = glob.glob('../[A,B]x[A,B]/methylDackel_cytosine_report/F*_1*cytosine_report.txt') # This is the path to the MethylDackel files generated by the cytosine report argument.\n",
    "\n",
    "shared = np.loadtxt('../SHARED_CpGs.Brahman_coords.txt')\n",
    "shared_tuple = tuple(shared)\n",
    "shared_cpgs = pd.read_csv('../../between_ref_SNPs/Angus.CpGs.to.Brahman_coords.SHARED.bed',\n",
    "                          header=None,\n",
    "                          index_col=None,\n",
    "                          sep='\\t')\n",
    "shared = []\n",
    "for row in shared_cpgs.itertuples():\n",
    "    shared.append(f'{row[1]}.{row[2]+1}')\n",
    "    shared.append(f'{row[1]}.{row[3]}')\n",
    "shared = tuple(shared)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in tqdm(methylkit_files):\n",
    "    group = file.split('/')[1]\n",
    "    df = pd.read_csv(file,\n",
    "            header=None,\n",
    "            sep='\\t',\n",
    "            names=['chromosome','base','strand', 'meth_C_count','unmeth_C_count','C_context','trinucleotide_context'],\n",
    "            dtype = {0:str,\n",
    "                    1:int,\n",
    "                    2:str,\n",
    "                    3:int,\n",
    "                    4:int,\n",
    "                    5:str,\n",
    "                    6:str})\n",
    "\n",
    "    chrbase = [f'{row.chromosome}.{row.base}' for row in df.itertuples()]\n",
    "    #df['chrBase'] = chrbase\n",
    "\n",
    "    name = os.path.basename(file)\n",
    "    name = name.split('.')[0]\n",
    "    df_dict = {}\n",
    "    df_dict['chrBase'] = chrbase\n",
    "    df_dict['chromosome'] = df['chromosome'].values\n",
    "    df_dict['base'] = df['base'].values\n",
    "    df_dict['strand'] = df['strand'].values\n",
    "    df_dict['context'] = ['CpG' for i in range(df.shape[0])]\n",
    "    proportion_c = np.array(df['meth_C_count'] / (df['meth_C_count'] + df['unmeth_C_count']))\n",
    "    proportion_c[np.isnan(proportion_c)] = 0.\n",
    "    df_dict['proportion_C'] = proportion_c\n",
    "    df_dict['coverage'] = df['meth_C_count'] + df['unmeth_C_count']\n",
    "\n",
    "    new_df = pd.DataFrame.from_dict(df_dict)\n",
    "    filtered_df = new_df[new_df['chrBase'].isin(shared)]\n",
    "    new_df = new_df.iloc[:, 1:]\n",
    "    filtered_df = filtered_df.iloc[:, 1:]\n",
    "    filtered_df.to_csv(f'./aligned2Brahman/{group}/{name}.Consensus.CpGs.meth',\n",
    "              header=None,\n",
    "              index_label=None,\n",
    "              index=None,\n",
    "              sep='\\t')\n",
    "    new_df.to_csv(f'./aligned2Brahman/{group}/{name}.CpGs.meth',\n",
    "              header=None,\n",
    "              index_label=None,\n",
    "              index=None,\n",
    "              sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "methylation_dl",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
